File: packages/js/src/boundary.ts
import { TAU } from "./constants";
import type { Bake3DResult, BoundaryOptions, BoundedField, FlowField, Out6, Sdf, Vec3 } from "./types";

// Bridson's free-slip ramp (the curl-noise boundary quintic): r(0) = 0 but r'(0) = 15/8 > 0, so
// the wall value is a pure tangential slip flow rather than a no-slip dead zone; r(1) = 1 with
// r'(1) = r''(1) = 0, so the field blends C²-smoothly into the unconstrained one.
function ramp(x: number): number {
  if (x <= 0) return 0;
  if (x >= 1) return 1;
  const x2 = x * x;
  return (x * (15 - 10 * x2 + 3 * x2 * x2)) / 8;
}
function dramp(x: number): number {
  if (x < 0 || x >= 1) return 0;
  const w = 1 - x * x;
  return (15 / 8) * w * w;
}

/**
 * A field constrained by an SDF obstacle. Velocity is `∇×(ramp(d/th)·A)` with `A` the base
 * field's analytic vector potential, expanded exactly as
 * `u_b = ramp'·∇d×A + ramp·u`
 * — divergence-free by construction (it is a curl), tangent to the wall (the normal component
 * carries a factor ramp(0) = 0), zero inside, and bit-identical to the base field beyond the
 * influence band.
 */
export class BoundedFieldImpl implements BoundedField {
  readonly base: FlowField;
  readonly sdf: Sdf;
  private th: number;
  private h: number;
  private grad?: (x: number, y: number, z: number) => ArrayLike<number>;
  private _ua: number[] = [0, 0, 0, 0, 0, 0];
  private _fa: number[] = [0, 0, 0];
  private _fb: number[] = [0, 0, 0];

  constructor(base: FlowField, sdf: Sdf, opts?: BoundaryOptions) {
    this.base = base;
    this.sdf = sdf;
    this.th = Math.max(opts?.thickness ?? 1, 1e-9);
    this.h = opts?.fdStep ?? 1e-3;
    this.grad = opts?.gradient;
  }

  /** Core: bounded velocity into out[0..2]. */
  private _u(x: number, y: number, z: number, t: number, out: number[] | Out6, o = 0): void {
    const d = this.sdf(x, y, z);
    if (d <= 0) { out[o] = 0; out[o + 1] = 0; out[o + 2] = 0; return; }
    const ua = this._ua;
    this.base.sampleUA(x, y, z, ua, t);
    const q = d / this.th;
    if (q >= 1) { out[o] = ua[0]; out[o + 1] = ua[1]; out[o + 2] = ua[2]; return; }
    let gx: number, gy: number, gz: number;
    if (this.grad) {
      const g = this.grad(x, y, z);
      gx = g[0]; gy = g[1]; gz = g[2];
    } else {
      const h = this.h, s = this.sdf;
      gx = (s(x + h, y, z) - s(x - h, y, z)) / (2 * h);
      gy = (s(x, y + h, z) - s(x, y - h, z)) / (2 * h);
      gz = (s(x, y, z + h) - s(x, y, z - h)) / (2 * h);
    }
    const r = ramp(q), rp = dramp(q) / this.th;
    const cx = gy * ua[5] - gz * ua[4]; // ∇d × A
    const cy = gz * ua[3] - gx * ua[5];
    const cz = gx * ua[4] - gy * ua[3];
    out[o] = rp * cx + r * ua[0];
    out[o + 1] = rp * cy + r * ua[1];
    out[o + 2] = rp * cz + r * ua[2];
  }

  sample(x: number, y: number, z: number, t = 0): Vec3 {
    const f = this._fa;
    this._u(x, y, z, t, f);
    return [f[0], f[1], f[2]];
  }

  sampleUW<T extends Out6>(x: number, y: number, z: number, out6: T, t = 0): T {
    this._u(x, y, z, t, out6, 0);
    const h = this.h, a = this._fa, b = this._fb;
    this._u(x, y + h, z, t, a); this._u(x, y - h, z, t, b);
    const uzy = (a[2] - b[2]) / (2 * h), uxy = (a[0] - b[0]) / (2 * h);
    this._u(x, y, z + h, t, a); this._u(x, y, z - h, t, b);
    const uyz = (a[1] - b[1]) / (2 * h), uxz = (a[0] - b[0]) / (2 * h);
    this._u(x + h, y, z, t, a); this._u(x - h, y, z, t, b);
    const uyx = (a[1] - b[1]) / (2 * h), uzx = (a[2] - b[2]) / (2 * h);
    out6[3] = uzy - uyz;
    out6[4] = uxz - uzx;
    out6[5] = uyx - uxy;
    return out6;
  }

  vorticity(x: number, y: number, z: number, t = 0): Vec3 {
    const o = this._ua;
    this.sampleUW(x, y, z, o, t);
    return [o[3], o[4], o[5]];
  }

  helicityDensity(x: number, y: number, z: number, t = 0): number {
    const o = this._ua;
    this.sampleUW(x, y, z, o, t);
    return o[0] * o[3] + o[1] * o[4] + o[2] * o[5];
  }

  potential(x: number, y: number, z: number, t = 0): Vec3 {
    const d = this.sdf(x, y, z);
    if (d <= 0) return [0, 0, 0];
    const ua = this._ua;
    this.base.sampleUA(x, y, z, ua, t);
    const r = ramp(d / this.th);
    return [r * ua[3], r * ua[4], r * ua[5]];
  }

  bake3D(n: number, t = 0): Bake3DResult {
    const data = new Float32Array(n * n * n * 4), o = [0, 0, 0, 0, 0, 0];
    let p = 0;
    for (let z = 0; z < n; z++) for (let y = 0; y < n; y++) for (let x = 0; x < n; x++) {
      this.sampleUW((x / n) * TAU, (y / n) * TAU, (z / n) * TAU, o, t);
      data[p] = o[0]; data[p + 1] = o[1]; data[p + 2] = o[2];
      data[p + 3] = o[0] * o[3] + o[1] * o[4] + o[2] * o[5];
      p += 4;
    }
    return { data, size: n, channels: 4 };
  }

  bakePotential3D(n: number, t = 0): Bake3DResult {
    const data = new Float32Array(n * n * n * 4), ua = this._ua;
    let p = 0;
    for (let z = 0; z < n; z++) for (let y = 0; y < n; y++) for (let x = 0; x < n; x++) {
      const px = (x / n) * TAU, py = (y / n) * TAU, pz = (z / n) * TAU;
      const d = this.sdf(px, py, pz);
      if (d <= 0) {
        data[p] = 0; data[p + 1] = 0; data[p + 2] = 0;
      } else {
        this.base.sampleUA(px, py, pz, ua, t);
        const r = ramp(d / this.th);
        data[p] = r * ua[3]; data[p + 1] = r * ua[4]; data[p + 2] = r * ua[5];
      }
      data[p + 3] = d;
      p += 4;
    }
    return { data, size: n, channels: 4 };
  }
}

File: packages/rust/src/boundary.rs
//! Free-slip SDF obstacle boundary.
//!
//! Given a signed-distance function `d = sdf(x, y, z)` (negative inside the obstacle) and the
//! base field's analytic vector potential `A`, the bounded velocity is the curl of a ramped
//! potential, expanded exactly as `u_b = ramp'(d/th) * (grad(d) x A) + ramp(d/th) * u`. It is
//! divergence-free by construction, tangent to the wall, zero inside, and identical to the base
//! field beyond the influence band.

use crate::field::HelixField;

/// Options for [`HelixField::with_boundary`](crate::HelixField::with_boundary).
pub struct BoundaryOptions {
    /// Width of the influence band, in world units. Clamped to `>= 1e-9`.
    pub thickness: f64,
    /// Finite-difference step for numerical gradients (SDF gradient and bounded vorticity).
    pub fd_step: f64,
    /// Optional analytic SDF gradient `grad(d)`. When absent, central differences are used.
    pub gradient: Option<Box<dyn Fn(f64, f64, f64) -> [f64; 3]>>,
}

impl Default for BoundaryOptions {
    fn default() -> Self {
        BoundaryOptions {
            thickness: 1.0,
            fd_step: 1e-3,
            gradient: None,
        }
    }
}

/// Bridson's free-slip ramp (the curl-noise boundary quintic). `r(0) = 0` but `r'(0) = 15/8`,
/// so the wall value is a pure tangential slip flow. `r(1) = 1` with `r'(1) = r''(1) = 0`.
#[inline]
fn ramp(x: f64) -> f64 {
    if x <= 0.0 {
        return 0.0;
    }
    if x >= 1.0 {
        return 1.0;
    }
    let x2 = x * x;
    (x * (15.0 - 10.0 * x2 + 3.0 * x2 * x2)) / 8.0
}

#[inline]
fn dramp(x: f64) -> f64 {
    if x < 0.0 || x >= 1.0 {
        return 0.0;
    }
    let w = 1.0 - x * x;
    (15.0 / 8.0) * w * w
}

/// A base [`HelixField`] constrained by an SDF obstacle. Borrows the base field.
pub struct BoundedField<'f, S>
where
    S: Fn(f64, f64, f64) -> f64,
{
    base: &'f HelixField,
    sdf: S,
    th: f64,
    h: f64,
    grad: Option<Box<dyn Fn(f64, f64, f64) -> [f64; 3]>>,
}

impl<'f, S> BoundedField<'f, S>
where
    S: Fn(f64, f64, f64) -> f64,
{
    pub(crate) fn new(base: &'f HelixField, sdf: S, opts: BoundaryOptions) -> Self {
        BoundedField {
            base,
            sdf,
            th: opts.thickness.max(1e-9),
            h: opts.fd_step,
            grad: opts.gradient,
        }
    }

    /// Bounded velocity at `(x, y, z, t)`.
    fn u(&self, x: f64, y: f64, z: f64, t: f64) -> [f64; 3] {
        let d = (self.sdf)(x, y, z);
        if d <= 0.0 {
            return [0.0, 0.0, 0.0];
        }
        let (u_base, a) = self.base.sample_ua(x, y, z, t);
        let q = d / self.th;
        if q >= 1.0 {
            return u_base;
        }
        let (gx, gy, gz);
        if let Some(g) = &self.grad {
            let gg = g(x, y, z);
            gx = gg[0];
            gy = gg[1];
            gz = gg[2];
        } else {
            let h = self.h;
            gx = ((self.sdf)(x + h, y, z) - (self.sdf)(x - h, y, z)) / (2.0 * h);
            gy = ((self.sdf)(x, y + h, z) - (self.sdf)(x, y - h, z)) / (2.0 * h);
            gz = ((self.sdf)(x, y, z + h) - (self.sdf)(x, y, z - h)) / (2.0 * h);
        }
        let r = ramp(q);
        let rp = dramp(q) / self.th;
        let cx = gy * a[2] - gz * a[1];
        let cy = gz * a[0] - gx * a[2];
        let cz = gx * a[1] - gy * a[0];
        [rp * cx + r * u_base[0], rp * cy + r * u_base[1], rp * cz + r * u_base[2]]
    }

    /// Bounded velocity at `(x, y, z, t)`.
    pub fn sample(&self, x: f64, y: f64, z: f64, t: f64) -> [f64; 3] {
        self.u(x, y, z, t)
    }

    /// Bounded velocity and its vorticity, the latter by central differences of the bounded
    /// velocity itself (`O(fd_step^2)`). Returns `(u, w)`.
    pub fn sample_uw(&self, x: f64, y: f64, z: f64, t: f64) -> ([f64; 3], [f64; 3]) {
        let u = self.u(x, y, z, t);
        let h = self.h;
        let ayp = self.u(x, y + h, z, t);
        let aym = self.u(x, y - h, z, t);
        let uzy = (ayp[2] - aym[2]) / (2.0 * h);
        let uxy = (ayp[0] - aym[0]) / (2.0 * h);
        let azp = self.u(x, y, z + h, t);
        let azm = self.u(x, y, z - h, t);
        let uyz = (azp[1] - azm[1]) / (2.0 * h);
        let uxz = (azp[0] - azm[0]) / (2.0 * h);
        let axp = self.u(x + h, y, z, t);
        let axm = self.u(x - h, y, z, t);
        let uyx = (axp[1] - axm[1]) / (2.0 * h);
        let uzx = (axp[2] - axm[2]) / (2.0 * h);
        let w = [uzy - uyz, uxz - uzx, uyx - uxy];
        (u, w)
    }

    /// Bounded vorticity at `(x, y, z, t)`.
    pub fn vorticity(&self, x: f64, y: f64, z: f64, t: f64) -> [f64; 3] {
        self.sample_uw(x, y, z, t).1
    }

    /// Bounded helicity density `u . w`.
    pub fn helicity_density(&self, x: f64, y: f64, z: f64, t: f64) -> f64 {
        let (u, w) = self.sample_uw(x, y, z, t);
        u[0] * w[0] + u[1] * w[1] + u[2] * w[2]
    }

    /// Bounded vector potential: `ramp(d/th) * A_base`, or zero inside the obstacle.
    pub fn potential(&self, x: f64, y: f64, z: f64, t: f64) -> [f64; 3] {
        let d = (self.sdf)(x, y, z);
        if d <= 0.0 {
            return [0.0, 0.0, 0.0];
        }
        let (_, a) = self.base.sample_ua(x, y, z, t);
        let r = ramp(d / self.th);
        [r * a[0], r * a[1], r * a[2]]
    }
}
